---
title: "miR-375 KO (baseline)"
subtitle: "Part3: Cell type assignment"
author: "Amy Hung"
date: "Last compiled on `r format(Sys.time(), '%d %B %Y')`"
output:
  rmdformats::robobook:
    thumbnails: true
    lightbox: true
    gallery: true
    code_folding: show
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir='/home/yah6/2022_04_375KO-12m_reanalysis/rmarkdown')
```

## Sample description

This is the single cell RNAseq data of two 375KO mice and two WT mice (generated by 10X Genomics Chromium platform - 3' V2 kit). Samples were jejunal crypts isolated from sex and age matched 375KO and WT mice (male; 12m old; chow diet condition). Set 1 contained one 375KO and one WT mice processed in April 2019. Set 2 contained one 375KO and one WT mice processed in October 2019.

------------------------------------------------------------------------
## Load required libraries

```{r load-libraries, message = FALSE, warning= FALSE, error=FALSE}
library(Seurat, lib.loc = "/programs/R-4.0.5/library")
library(tidyverse, lib.loc = "/programs/R-4.0.5/library")
#library(MAST, lib.loc = "/home/yah6/R/x86_64-pc-linux-gnu-library/4.0")
```

## Find cluster markers

I used function `FindAllMarkers` to identify gene markers for each of the Seurat clusters. Test method MAST fits two-part, generalized linear models that are specially adapted for bimodal and/or zero-inflated single cell gene expression data.
*Notes*: For rendering purpose, you will see I directly load the marker gene list that I generated in April 2022.

```{r find-markers, warning=F, message=F}
#set.seed(1234)

# Load Seurat object
proj_name <- "375KO-12m"
seurat_obj <- readRDS(paste0(proj_name, '_seurat-obj.rds'))

# Make sure Idents are Seurat clusters 
Idents(seurat_obj) <- "seurat_clusters"
levels(seurat_obj)

# # Define markers for each cluster 
# markers <- FindAllMarkers(seurat_obj, 
#                           min.pct = -Inf,   
#                           logfc.threshold = .5, 
#                           only.pos = T, # only look for enriched genes, not depleted genes
#                           test.use = "MAST", 
#                           latent.vars = c('percent.mt', 'nFeature_RNA'))
# 
# # Write output
# markers <- markers %>%
#   filter(avg_log2FC > 0) %>%
#   write_csv(paste0(proj_name, '_HighlyEnrichedMarkers_logFC.5.csv'))
# head(markers)

markers <- read.csv('/home/yah6/2022_04_375KO-12m_reanalysis/375KO-12m_HighlyEnrichedMarkers_logFC.5.csv')
head(markers)
```
## Determine cell type 

### Simple overlapping method

`geneset_marker_overlap.f` is a custom function the checks the percent overlap between a marker gene list of a certain cell type with the markers of each cluster (as determined by the FindMarkers function).

```{r functions}
geneset_marker_overlap.f <- function(SerObj, enrichedGenes_df, known_markers, FC_thresh = 0, projname = 'proj', w = NULL, h = NULL, facet_ncol = 5, save_plot = T) {

  ### Loop through each cluster and calculate the percentage of genes from each cell type in that cluster
  #     above the FC_thresh using the output from finding highly enriched markers of each cluster
  overlap_df <- data.frame()
  for (i in sort(unique(Idents(SerObj)))) {
    cat(paste0('\nCalculating geneset percentages for cluster ', i, '...'))
    df_tmp <- enrichedGenes_df %>%
      filter(cluster == i) %>%
      filter(avg_log2FC > FC_thresh)

    for (t in unique(sort(known_markers$type))) {

      cell_type_markers <- known_markers %>%
        filter(type == t) %>%
        pull(gene) %>%
        unique() %>%
        as.vector()

      per <- length(intersect(df_tmp$gene, cell_type_markers)) / length(cell_type_markers) * 100
      genes <- intersect(df_tmp$gene, cell_type_markers)

      overlap_df <- bind_rows(overlap_df, data.frame('Cluster' = i,
                                                     'Type' = t,
                                                     'Percent' = round(per, 2),
                                                     'Overlap_genes' = paste(genes, collapse = '|'),
                                                     'number_of_cluster_markers' = length(df_tmp$gene)))
    }
  }
  cat('\n\n')
  
  ### Reformat the dataframe to be easier to read, write to csv, and return
  overlap_df %>%
    dplyr::select(Cell_Type = Type, everything()) %>%
    write_csv(paste0(projname, '_FC', FC_thresh, '.csv'))

  ### Refactor cluster so appear in numerical order
  overlap_df <- overlap_df %>%
    mutate(Cluster = factor(Cluster, levels = unique(Cluster)))

  ### Create a bar plot to show the percent cell type cells present in each cluster
  overlap_df <- overlap_df %>%
    mutate(facet_name = paste0(Cluster, " (", number_of_cluster_markers, ")"))

  overlap_df$facet_name <- reorder(overlap_df$facet_name, as.numeric(overlap_df$Cluster))

  ### Plot
  g <- ggplot(overlap_df, aes(Type, Percent, alpha = Percent)) +
          geom_bar(stat = 'identity', color = 'black') +
          facet_wrap(~facet_name, ncol = facet_ncol, scales = 'free_y') +
          scale_x_discrete(NULL) +
          theme_bw() +
          theme(text = element_text(size = 20),
                axis.title = element_text(size = 18, face = 'bold'),
                axis.text.y = element_text(size = 16, color = 'black'),
                axis.text.x = element_text(size = 16, color = 'black', angle = 90, hjust = 1, vjust = .5),
                strip.text = element_text(face = 'bold'),
                panel.grid = element_blank())
  
  if (save_plot == T) {
      filename = paste0(projname, '_FC', FC_thresh, '.png')
      print(paste0('Saving percent image as: ', filename))

  ### Determine width and height
  if (is.null(w)) {
    w = min(10, length(unique(overlap_df$Cluster)) * 2 + 3)
  }

  if (is.null(h)) {
    h = max(3, 2 * ceiling(length(unique(overlap_df$Cluster)) / 4))
  }
      
      g
      ggsave(filename, width = w, height = h)
      
  }


  return(g)
}
```

I used Haber et al., Nature (2017) as the "known cell type marker" and overlap it with the cluster markers defined in my dataset. Be noted that the previous analyses of this dataset revealed a cluster with "unknown cell type". We later found that this "unknown" cluster represented intraepithelial lymphocytes (IEL) associated with the intestinal tract, as it was featured by an immune cell signature. In addition to Haber gene list for defining intestinal epithelial cell types, we used a custom gene list of IEL for calling IEL cluster. 

```{r cell-type-markers, fig.width=14, fig.height=10}
dir.create("cell_type_assignment")
cell_marker_dir <- "./"

## Haber
cell_markers <- read_csv("/home/yah6/2022_04_375KO-12m_reanalysis/haber_assignment_list.csv")

geneset_marker_overlap.f(seurat_obj,
                         markers,
                         cell_markers,
                         FC_thresh = .5, 
                         projname = paste0("cell_type_assignment/", proj_name, '_CTA_Haber'))

# Non-core scRNA analysis code, specifically to check for IEL
iel <- data.frame(type = "IEL", gene = c("Gzma", "Gzmb", "Itgae", "Ccl5", "Cd7", "Cd69", "Cd3g", "Cd8a"))

geneset_marker_overlap.f(seurat_obj,
                         markers,
                         iel,
                         FC_thresh = .5, 
                         projname = paste0("cell_type_assignment/", proj_name, '_CTA_IEL'))


```


### Cumulative expression of cell type marker list across clusters 

For a list of cell type markers, calculate a expression score for each cluster using `AddModuleScore` function. An in-depth explaination of what is happening behind the scene can be found in a [blog post by Walter Muskovic](https://www.waltermuskovic.com/2021/04/15/seurat-s-addmodulescore-function/). We use `DotPlot` for data visualization. 

```{r module-expression, warning=F}
# create a temporary Seurat object here to avoid over populating our metadata with these scores
seurat_tmp <- seurat_obj


## DotPlot of Haber list
cell_markers <- read_csv("/home/yah6/2022_04_375KO-12m_reanalysis/haber_assignment_list.csv")

goi_li <- split(cell_markers$gene, cell_markers$type)
seurat_tmp <- AddModuleScore(object = seurat_tmp, features = goi_li, name = "MarkerList")

DotPlot(seurat_tmp, features = paste0("MarkerList", seq(1, length(goi_li)))) +  
  scale_x_discrete(NULL,
                   labels = names(goi_li)) +
  theme(axis.text.x = element_text(angle = 90, vjust = .5, hjust = 1))

ggsave(paste0("cell_type_assignment/", proj_name, "_DP-CTA_Haber.png"), units = "in", width = 2 + length(goi_li) * 1, height = 8)


## DotPlot of IEL list
cell_markers <- data.frame(type = "IEL", gene = c("Gzma", "Gzmb", "Itgae", "Ccl5", "Cd7", "Cd69", "Cd3g", "Cd8a"))

goi_li <- split(cell_markers$gene, cell_markers$type)
seurat_tmp <- AddModuleScore(object = seurat_tmp, features = goi_li, name = "MarkerList")

DotPlot(seurat_tmp, features = paste0("MarkerList", seq(1, length(goi_li)))) + 
  scale_x_discrete(NULL,
                   labels = names(goi_li)) +
  theme(axis.text.x = element_text(angle = 90, vjust = .5, hjust = 1))

ggsave(paste0("cell_type_assignment/", proj_name, "_DP-CTA_IEL.png"), units = "in", width = 2 + length(goi_li) * 1.2, height = 8)


```

## Assign cell types

Based on the results of method A (overlap analysis) and method B (the module expression scores), we then assign cell types. This is curated manually and saved as a <name>_cluster_assignment.csv file. This code chunk also writes RDS file that includes "cell type information" for downstream analyses.  Be noted that I save two RDS files for this dataset - one that includes ALL cells and the other includes only epithelial cells.   

```{r cell-type-assignment, fig.width=5, fig.height=5}

cluster_ass <- read_csv('/home/yah6/2022_04_375KO-12m_reanalysis/375KO-12m_cluster_assignment.csv') %>%
  mutate(cluster_assignment = paste(cluster, assignment, sep = "_"))

seurat_obj[["cluster_id_named"]] <- plyr::mapvalues(Idents(seurat_obj),
                                                    from=cluster_ass$cluster,
                                                    to=cluster_ass$cluster_assignment)

new.cluster.ids <- cluster_ass$assignment
names(new.cluster.ids) <- levels(seurat_obj)
seurat_obj <- RenameIdents(seurat_obj, new.cluster.ids)

seurat_obj[["cluster_named"]] <- Idents(seurat_obj)


## All cells
DimPlot(seurat_obj, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
ggsave(paste0(proj_name, "_umap_clusterNamed.png"), units = 'in', width = 6, height = 6, dpi = 250)

# saveRDS(seurat_obj, paste0(proj_name, 'clusteredName_seurat-obj.rds'))


## Epithelial cell only (aka excluding IEL cells)
seurat_obj_noIEL <- subset(seurat_obj, subset = cluster_named != "IEL")

DimPlot(seurat_obj_noIEL, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
ggsave(paste0(proj_name, "_umap_clusterNamed_noIEL.png"), units = 'in', width = 6, height = 6, dpi = 250)

# saveRDS(seurat_obj_noIEL, paste0(proj_name, 'clusteredName_noIEL_seurat-obj.rds')) 

```

## Find markers of named clusters

```{r find-markers-named, eval = F, warning=F}
markers <- FindAllMarkers(seurat_obj, 
                          min.pct = -Inf, 
                          logfc.threshold = .25, 
                          only.pos = T, 
                          test.use = "MAST",
                          latent.vars = c('percent.mt', 'nFeature_RNA'))

markers <- markers %>%
  filter(avg_log2FC > 0) %>%
  write_csv(paste0(proj_name, '_HighlyEnrichedMarkers_clusteredName_logFC.25.csv'))
```

## Session info

```{r session-info}

sessionInfo()

```
