---
title: "miR-375 KO (baseline)"
subtitle: "Milo: Differential abundance analysis (test)"
author: "Amy Hung"
date: "Last compiled on `r format(Sys.time(), '%d %B %Y')`"
output:
  rmdformats::robobook:
    thumbnails: true
    lightbox: true
    gallery: true
    code_folding: show
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir='/home/yah6/2022_04_375KO-12m_reanalysis/rmarkdown')
```

## Sample description

This is the single cell RNAseq data of two 375KO mice and two WT mice (generated by 10X Genomics Chromium platform - 3' V2 kit). Samples were jejunal crypts isolated from sex and age matched 375KO and WT mice (male; 12m old; chow diet condition). Set 1 contained one 375KO and one WT mice processed in April 2019. Set 2 contained one 375KO and one WT mice processed in October 2019.

------------------------------------------------------------------------

## Load required libraries
```{r load-libraries, message = FALSE, warning= FALSE, error=FALSE}
library(stringr)
library(Seurat)
library(tidyverse)

library(BiocParallel)
library(glmGamPoi)

library(scater)
library(miloR)

```

## Milo (package miloR) 
Tool Milo was developed by [Dann et al., Nature Biotechnoloty (2021)](https://www.nature.com/articles/s41587-021-01033-z). It represents a scalable statistical framework that performs differential abundance testing by assigning cells to partially overlapping neighborhoods on a k-nearest neighbor graph. It does not require prior knowledge of cell type of your scRNA-seq. I basically followed this [vignette](https://bioconductor.riken.jp/packages/devel/bioc/vignettes/miloR/inst/doc/milo_demo.html) to test this method with my dataset. 

## Data preparation
The current version of miloR has issues directly working on a Seurat object with pre-calculated clustering. If one wants to present miloR result by overlay it in a UMAP, he/shes needs to repeat Seurat clustering step.  

## Make Milo object
Load the Seurat object with pre-calculated clusteirng info. Here I directly load Milo object I made previoulsy.

``` {r Seurat-object}
# load Seurat object (done with doublet removal, filtering and clustering)
proj_name <- "375KO-12m"
seurat_obj <- readRDS("/home/yah6/2022_04_375KO-12m_reanalysis/375KO-12mclusteredName_seurat-obj.rds")
```

```{r Milo-object, eval = F}
# convert to Milo object
seurat_obj_sce <- as.SingleCellExperiment(seurat_obj)
seurat_obj_milo <- Milo(seurat_obj_sce)

# save Milo object 
saveRDS(seurat_obj_milo, paste0(proj_name, '_seurat-obj_milo.rds'))
```

## Construct KNN graph and define neighbors
After defining representative neighborhoods, check out using `plotNhoodSizeHist` function. Empirically, we found itâ€™s best to have a distribution peaking between 50 and 100. Otherwise you might consider rerunning makeNhoods increasing k and/or prop. 
*Note*: My dataset is small so it doesn't not look great.

```{r define-neighborhoods, warning=F}

seurat_obj_milo <- readRDS("/home/yah6/2022_04_375KO-12m_reanalysis/375KO-12m_seurat-obj_milo.rds")

# Construct KNN graph
seurat_obj_milo <- buildGraph(seurat_obj_milo, k = 20, d = 30)

# Define representative neighborhoods
seurat_obj_milo <- makeNhoods(seurat_obj_milo, prop = 0.1, k = 20, d=30, refined = TRUE)

# Check neighborhoods 
plotNhoodSizeHist(seurat_obj_milo)
ggsave(paste0(proj_name, "_MiloR_NhoodSizeHist_k20.png"), units = "in", width = 5, height = 4, dpi = 300)
```

## Count cells in neighborhoods
Count how many cells from each sample are in each neighborhood. This count matrix will be used for DA testing.
```{r count-cells, warning=F}
meta.df <- data.frame(SampleID = seurat_obj_milo@colData$mouse, 
                      Condition = seurat_obj_milo@colData$group)
seurat_obj_milo_count <- countCells(seurat_obj_milo, meta.data = meta.df, sample="SampleID")

head(nhoodCounts(seurat_obj_milo_count))
```

## Differential Abundence test
DA test is based on Negative Binomial GLM implementation in `edgeR`. For computting neighborhood connectivity, Milo uses an adaptation of the Spatial FDR correction introduced by `cydar`.
*Note*: Since my dataset is small, nothing really came out as significant.

```{r DA-test, warning=F, message = FALSE}
# Make design matrix
milo_design <- meta.df[,c("SampleID", "Condition")]
milo_design <- distinct(milo_design)
milo_design

# Design matrix matches samples to a condition of interest. 
rownames(milo_design) <- milo_design$SampleID

# Compute neighbourhood connectivity
dat_milo <- calcNhoodDistance(seurat_obj_milo_count, d=30)

# Perform DA analysis
dat_results <- testNhoods(dat_milo, design = ~ Condition, design.df = milo_design)

# Check DA result
dat_results %>%
  arrange(SpatialFDR) %>%
  head()

# QC plots 
ggplot(dat_results, aes(PValue)) + geom_histogram(bins=50)

ggplot(dat_results, aes(logFC, -log10(SpatialFDR))) + 
  geom_point() +
  geom_hline(yintercept = 1) ## Mark significance threshold (10% FDR)
```

## Visualization 
*Note*: since there's no significant DA in my dataset, I also ploted using FDR threshold > 0.999 to show fold change difference between WT and KO. 

```{r visualization, warning=F, fig.width=8, fig.height=6, message = FALSE}
dat_milo <- buildNhoodGraph(dat_milo)

plotNhoodGraphDA(dat_milo, dat_results, alpha = 0.05) 
ggsave(paste0(proj_name, "_MiloR_k20_plotNhoodGraphDA_FDR0.05_test.png"))

plotNhoodGraphDA(dat_milo, dat_results, alpha = 0.999) 
ggsave(paste0(proj_name, "_MiloR_k20_plotNhoodGraphDA_FDR0.999_test.png"))
```




