---
title: "miR-375 KO (baseline)"
subtitle: "Part1: Ambient RNA removal"
author: "Amy Hung"
date: "Last compiled on `r format(Sys.time(), '%d %B %Y')`"
output:
  rmdformats::robobook:
    thumbnails: true
    lightbox: true
    gallery: true
    code_folding: show
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir='/home/yah6/2022_04_375KO-12m_reanalysis/rmarkdown')
```

## Sample description

This is the single cell RNAseq data of two 375KO mice and two WT mice (generated by 10X Genomics Chromium platform - 3' V2 kit). Samples were jejunal crypts isolated from sex and age matched 375KO and WT mice (male; 12m old; chow diet condition).

------------------------------------------------------------------------

## Load required libraries

```{r load-libraries, message = FALSE, warning= FALSE, error=FALSE}
library(Seurat, lib.loc = "/programs/R-4.0.5/library")
library(tidyverse, lib.loc = "/programs/R-4.0.5/library")
library(SoupX)
#library(DropletUtils, lib.loc = "/home/yah6/R/x86_64-pc-linux-gnu-library/4.0")
```

## Demultiplexing

The demultiplexing of these samples was done by Cornell core.

## Alignment and quantification

Alignment and quantification of single cell data is done using cellranger count from the 10X genomics cellranger software package.

Cellranger count performs alignment, filtering, barcode counting, and UMI counting. It uses the Chromium cellular barcodes to generate feature-barcode matrices, determine clusters, and perform gene expression analysis.

Expanding on the general overview, demultiplexed fastqs are aligned to the genome using STAR. Reads aligning to the transcriptome are grouped by barcode, UMI, and associated gene. If the same barcode and UMI are associated with mulitple genes, the more supported gene group is kept and the other is discarded. After filtering, each barcode-UMI-gene is recorded as a single UMI count in the barcode-feature matrix. Next, cellranger calls cells based on the EmptyDrops method from Lun et al., 2018. This first calls high UMI content barcodes as cells. Secondly, it constructs a background model of the lowliest UMI-containing barcodes (empty droplets) and compares the profile of the remaining uncalled barcodes to the model. Barcodes poorly fittiing the background model are called as cells.

More information on cellranger can be found here: <https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome>

More information on the underlying cellranger algorithms can be found here: <https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview>

The cellranger count commands are based on the Broad Core command and is as follows:

```{r cellranger count code, eval=F}

### 375KO sample

#!/bin/bash
# KO from set 1
/programs/cellranger-3.0.1/cellranger count \
--id=KO-12m_v2 \
--sample=108168_H7YCKBGXC_2_KO \
--localcores=22 \
--fastqs=H2Y5TBGXB_Sethupathy_10414279_cellranger_mkfastq_output_20April19/outs/fastq_path/ \
--transcriptome=/home/pr46_0001/projects/genome/refdata-cellranger-mm10-1.2.0

# KO from set 2
/programs/cellranger-3.0.1/cellranger count \
--id=375KO_KO_2019oct \
--sample=95294_H2Y5TBGXB_KO \
--localcores=22 \
--fastqs=H7YCKBGXC_Singh_10422070_cellranger_mkfastq_output_15Oct19/outs/fastq_path/ \
--transcriptome=/home/pr46_0001/projects/genome/refdata-cellranger-mm10-1.2.0

### WT sample

#!/bin/bash
# WT from set 1
/programs/cellranger-3.0.1/cellranger count \
--id=WT-12m_v2 \
--sample=95294_H2Y5TBGXB_WT \
--localcores=22 \
--fastqs=H2Y5TBGXB_Sethupathy_10414279_cellranger_mkfastq_output_20April19/outs/fastq_path/ \
--transcriptome=/home/pr46_0001/projects/genome/refdata-cellranger-mm10-1.2.0

# WT from set 2
/programs/cellranger-3.0.1/cellranger count \
--id=375KO_WT_2019oct \
--sample=108167_H7YCKBGXC_1_WT \
--localcores=22 \
--fastqs=H7YCKBGXC_Singh_10422070_cellranger_mkfastq_output_15Oct19/outs/fastq_path/ \
--transcriptome=/home/pr46_0001/projects/genome/refdata-cellranger-mm10-1.2.0
```

Logs from these runs can be found on the Cornell HPC in the following directory: `/home/pr46_0001/scRNAseq/2019_10_375KO_12m/` and `/home/pr46_0001/scRNAseq/2019_04_375KO_12m/`.

## Removal of ambient RNA using SoupX - function

In single cell experiments, non-endogenous RNA can make up anywhere from 2-50% of all counts, although 10% seems like a decent expectation. To remove these exogenous counts, we can use the SoupX program that models the background expression from empty droplets. More information can be found in the [SoupX manuscript](https://academic.oup.com/gigascience/article/9/12/giaa151/6049831) or the [SoupX tutorial](https://rawcdn.githack.com/constantAmateur/SoupX/204b602418df12e9fdb4b68775a8b486c6504fe4/inst/doc/pbmcTutorial.html). I further took the advice of this [CellGen vignette](https://cellgeni.github.io/notebooks/html/new-10kPBMC-SoupX.html) to use the Seurat clustering. Other options for ambient RNA removal include Cell Bender, DecontX, and SoupOrCell.

```{r SoupX-function}
# Since we will be repeating SoupX on each of our 4 samples, I'll combine the code into a function
#
# The matrix files are output by cellranger
#     filtered = only has droplets with putative cell
#     raw = containts all droplets, both putative cells and empty droplets

SoupX.f <- function(raw.matrix, filtered.matrix) {
  # Create soup channel with the raw and filtered matrix
  soup.channel  <- SoupChannel(raw.matrix, filtered.matrix)
  print(soup.channel)

  # SoupX requires clusters in order to define marker genes. One can either use CellRanger clustering or quickly cluster using Seurat.
  # I chose to use Seurat here.

  seurat_obj <- CreateSeuratObject(counts = filtered.matrix)
  seurat_obj    <- SCTransform(seurat_obj, verbose = F)
  seurat_obj    <- RunPCA(seurat_obj, verbose = F)
  seurat_obj    <- RunUMAP(seurat_obj, dims = 1:30, verbose = F)
  seurat_obj    <- FindNeighbors(seurat_obj, dims = 1:30, verbose = F)
  seurat_obj    <- FindClusters(seurat_obj, verbose = T)

  # The cluster information from Seurat can be fed into the SoupX setClusters function
  meta <- seurat_obj@meta.data
  umap <- seurat_obj@reductions$umap@cell.embeddings
  soup.channel <- setClusters(soup.channel, setNames(meta$seurat_clusters, rownames(meta)))
  soup.channel <- setDR(soup.channel, umap)

  # After defining the clusters, we can run calculate the ambient RNA profile
  soup.channel <- autoEstCont(soup.channel)
  print(head(soup.channel$soupProfile[order(soup.channel$soupProfile$est, decreasing = T), ], n = 20))

  # We can adjust the counts and save the new count matrix
  adj.matrix <- adjustCounts(soup.channel, roundToInt = T)
  return(adj.matrix)
}
```

## Removal of ambient RNA with SoupX {.tabset}

### KO set 1

```{r SoupX-adjusted-counts-KO2, warning=F}

filtered.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_04_375KO_12m/KO-12m_v2/outs/filtered_gene_bc_matrices_h5.h5",use.names = T)
raw.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_04_375KO_12m/KO-12m_v2/outs/raw_gene_bc_matrices_h5.h5",use.names = T)
adj.counts <- SoupX.f(raw.matrix, filtered.matrix)

# write output (generated already)
#DropletUtils:::write10xCounts("/workdir/yah6/soupx_KO_filt", adj.counts, overwrite = T)

```

### KO set 2

```{r SoupX-adjusted-counts-KO1, warning=F}

filtered.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_10_375KO_12m/375KO_KO_2019oct/outs/filtered_feature_bc_matrix.h5",use.names = T)
raw.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_10_375KO_12m/375KO_KO_2019oct/outs/raw_feature_bc_matrix.h5",use.names = T)
adj.counts <- SoupX.f(raw.matrix, filtered.matrix)

# write output (generated already)
# DropletUtils:::write10xCounts("/workdir/yah6/soupx_KO2_filt", adj.counts, overwrite = T)

```

### WT set 1

```{r SoupX-adjusted-counts-WT2, warning=F}

filtered.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_04_375KO_12m/WT-12m_v2/outs/filtered_gene_bc_matrices_h5.h5",use.names = T)
raw.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_04_375KO_12m/WT-12m_v2/outs/raw_gene_bc_matrices_h5.h5",use.names = T)
adj.counts <- SoupX.f(raw.matrix, filtered.matrix)

# write output (generated already)
#DropletUtils:::write10xCounts("/workdir/yah6/soupx_WT1_filt", adj.counts, overwrite = T)

```

### WT set 2

```{r SoupX-adjusted-counts-WT1, warning=F}

filtered.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_10_375KO_12m/375KO_WT_2019oct/outs/filtered_feature_bc_matrix.h5",use.names = T)
raw.matrix <- Read10X_h5("/home/pr46_0001/scRNAseq/2019_10_375KO_12m/375KO_WT_2019oct/outs/raw_feature_bc_matrix.h5",use.names = T)
adj.counts <- SoupX.f(raw.matrix, filtered.matrix)

# write output (generated already)
#DropletUtils:::write10xCounts("/workdir/yah6/soupx_WT2_filt", adj.counts, overwrite = T)

```

------------------------------------------------------------------------

## Session info

```{r session-info}
sessionInfo()
```
