---
title: "miR-375 KO (baseline)"
subtitle: "Part4: Differential expression analysis"
author: "Amy Hung"
date: "Last compiled on `r format(Sys.time(), '%d %B %Y')`"
output:
  rmdformats::robobook:
    thumbnails: true
    lightbox: true
    gallery: true
    code_folding: show
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir='/home/yah6/2022_04_375KO-12m_reanalysis/rmarkdown')
```

## Sample description

This is the single cell RNAseq data of two 375KO mice and two WT mice (generated by 10X Genomics Chromium platform - 3' V2 kit). Samples were jejunal crypts isolated from sex and age matched 375KO and WT mice (male; 12m old; chow diet condition). Set 1 contained one 375KO and one WT mice processed in April 2019. Set 2 contained one 375KO and one WT mice processed in October 2019.

------------------------------------------------------------------------

## Load required libraries

```{r load-libraries, message = FALSE, warning= FALSE, error=FALSE, eval = F}
library(Seurat)
library(tidyverse, lib.loc = "/programs/R-4.0.5/library")
library(MAST)
```

## Differential expression within assigned cluster 

I use `FindMarkers` function for the differential expression. If one has multiple sequencing sets with potentially large batch effects, one could use `FindConservedMarkers` instead. Test method MAST identifies differentially expressed genes between two groups of cells using a hurdle model tailored to scRNA-seq data (Finak et al., 2015). For rendering purpose, I just load the DEG output that I generated previously.  

### Epithelial-cell-only dataset 

```{r load-data-noIEL}
seurat_obj <- readRDS('/home/yah6/2022_04_375KO-12m_reanalysis/375KO-12mclusteredName_noIEL_seurat-obj.rds')
seurat_obj
levels(seurat_obj)
```

```{r find-differentially-expressed-noIEL, warning=F, eval=F}
# create an empty dataframe to rbind DEG result
deg <- data.frame() 

# loop through all epithelial cell clusters
for (c in levels(Idents(seurat_obj))) {
  
  # print job
  print(paste0("Generating differential expression within cluster: ", c))
  sub_obj <- subset(seurat_obj, idents = c)
  print(levels(Idents(sub_obj)))
  Idents(sub_obj) <- "group"
  print(levels(Idents(sub_obj)))
  print(sub_obj)
  
  # find differentially expressed genes (DEG) 
  mks <- FindMarkers(sub_obj, 
                     ident.1 = '375KO', 
                     ident.2 = 'WT', 
                     test.use = "MAST", 
                     min.pct = -Inf, 
                     logfc.threshold = -Inf,
                     latent.vars = c('percent.mt', 'nFeature_RNA'))
  
  # calculate average expression level of WT and KO
  exp <- AverageExpression(sub_obj, assays = "RNA")$RNA %>%
    as.data.frame() %>%
    rownames_to_column(var = "gene")
  
  names(exp)[2:3] <- paste0("exp_", names(exp)[2:3])
  
  # combine DEG and average expression result
  mks <- mks %>% 
    as.data.frame() %>%
    mutate(gene = row.names(.),
           cluster = c) %>%
    left_join(exp)
  
  # paste to deg
  deg <- bind_rows(deg, mks)
  
}

# write output
write_csv(deg, file = paste0(proj_name, '_DEG_findMarkers_clusterNamed_375KOvsWT_noIEL.csv'))
```

```{r find-differentially-expressed-noIEL-deg, warning=F}
deg <- read.csv("/home/yah6/2022_04_375KO-12m_reanalysis/375KO-12m_DEG_findMarkers_clusterNamed_375KOvsWT.csv")
head(deg)
```


## Session info

```{r session-info}

sessionInfo()

```
